executor/foreign_key.go

```diff
@@ -643,51 +647,37 @@ func (fkc *FKCascadeExec) buildFKCascadePlan(ctx context.Context) (plannercore.P
 		return nil, nil
 	}
 	var indexName model.CIStr
-	indexForFK := model.FindIndexByColumns(fkc.childTable, fkc.fk.Cols...)
-	if indexForFK != nil {
-		indexName = indexForFK.Name
+	if fkc.fkIdx != nil {
+		indexName = fkc.fkIdx.Name
 	}
-	var sqlStr string
-	var err error
+	var stmtNode ast.StmtNode
 	switch fkc.tp {
 	case plannercore.FKCascadeOnDelete:
 		fkValues := fkc.fetchOnDeleteOrUpdateFKValues()
 		switch model.ReferOptionType(fkc.fk.OnDelete) {
 		case model.ReferOptionCascade:
-			sqlStr, err = GenCascadeDeleteSQL(fkc.referredFK.ChildSchema, fkc.childTable.Name, indexName, fkc.fk, fkValues)
+			stmtNode = GenCascadeDeleteAST(fkc.referredFK.ChildSchema, fkc.childTable.Name, indexName, fkc.fkCols, fkValues)
 		case model.ReferOptionSetNull:
-			sqlStr, err = GenCascadeSetNullSQL(fkc.referredFK.ChildSchema, fkc.childTable.Name, indexName, fkc.fk, fkValues)
+			stmtNode = GenCascadeSetNullAST(fkc.referredFK.ChildSchema, fkc.childTable.Name, indexName, fkc.fkCols, fkValues)
 		}
 	case plannercore.FKCascadeOnUpdate:
 		switch model.ReferOptionType(fkc.fk.OnUpdate) {
 		case model.ReferOptionCascade:
 			couple := fkc.fetchUpdatedValuesCouple()
 			if couple != nil && len(couple.NewValues) != 0 {
-				sqlStr, err = GenCascadeUpdateSQL(fkc.referredFK.ChildSchema, fkc.childTable.Name, indexName, fkc.fk, couple)
+				stmtNode = GenCascadeUpdateAST(fkc.referredFK.ChildSchema, fkc.childTable.Name, indexName, fkc.fkCols, couple)
 			}
 		case model.ReferOptionSetNull:
 			fkValues := fkc.fetchOnDeleteOrUpdateFKValues()
-			sqlStr, err = GenCascadeSetNullSQL(fkc.referredFK.ChildSchema, fkc.childTable.Name, indexName, fkc.fk, fkValues)
+			stmtNode = GenCascadeSetNullAST(fkc.referredFK.ChildSchema, fkc.childTable.Name, indexName, fkc.fkCols, fkValues)
 		}
 	}
-	if err != nil {
-		return nil, err
+	if stmtNode == nil {
+		return nil, errors.Errorf("generate foreign key cascade ast failed, %v", fkc.tp)
 	}
-	if sqlStr == "" {
-		return nil, errors.Errorf("generate foreign key cascade sql failed, %v", fkc.tp)
-	}
-
 	sctx := fkc.b.ctx
-	exec, ok := sctx.(sqlexec.RestrictedSQLExecutor)
-	if !ok {
-		return nil, nil
-	}
-	stmtNode, err := exec.ParseWithParams(ctx, sqlStr)
-	if err != nil {
-		return nil, err
-	}
 	ret := &plannercore.PreprocessorReturn{}
-	err = plannercore.Preprocess(ctx, sctx, stmtNode, plannercore.WithPreprocessorReturn(ret), plannercore.InitTxnContextProvider)
+	err := plannercore.Preprocess(ctx, sctx, stmtNode, plannercore.WithPreprocessorReturn(ret), plannercore.InitTxnContextProvider)
```

Comment:
What's the point of `plannercore.InitTxnContextProvider` here? Seems to me the only usage of it is handling state read which we don't want here?

Reply:
@winoros PTAL

Reply:
Look like we can remove `plannercore.WithPreprocessorReturn(ret), plannercore.InitTxnContextProvider`, @winoros